sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/io/AvroIO.java

```diff
@@ -326,6 +327,14 @@
       }
 
       @Override
+      public void populateDisplayData(DisplayData.Builder builder) {
+        builder.add("schema", type);
+        if (filepattern != null) {
+          builder.add("filePattern", filepattern);
+        }
```

Comment:
`.apply` is where all graph construction happens and it's the location we currently make all decisions like this. I don't think it's a great idea to break that abstraction, but you're right that there's a clear conflict.

I suspect the contract you'll end up wanting is one in which the display data is not valid until after `apply` has been called.